### ISSUE TITLE
all-in-one with badger crashes with nil pointer

### ISSUE THREAD

**Issue by surendarchandra** [jaegertracing/jaeger#1848]
## Requirement - what kind of business use case are you trying to solve?

Use all-in-one and badger until we deploy more robust storage backends

## Problem - what in Jaeger blocks you from solving the requirement?
Frequently, I get these traces (I suspect when someone views a span with many children, say 30,000):
2019/10/09 09:21:21 http: superfluous response.WriteHeader call from github.com/jaegertracing/jaeger/vendor/github.com/gorilla/handlers.recoveryHandler.ServeHTTP.func1 (recovery.go:73)
{"level":"error","ts":1570638081.307679,"caller":"recoveryhandler/zap.go:33","msg":"runtime error: invalid memory address or nil pointer dereference","stacktrace":"github.com/jaegertracing/jaeger/pkg/recoveryhandler.zapRecoveryWrapper.Println\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/pkg/recoveryhandler/zap.go:33\ngithub.com/jaegertracing/jaeger/vendor/github.com/gorilla/handlers.recoveryHandler.log\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/vendor/github.com/gorilla/handlers/recovery.go:83\ngithub.com/jaegertracing/jaeger/vendor/github.com/gorilla/handlers.recoveryHandler.ServeHTTP.func1\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/vendor/github.com/gorilla/handlers/recovery.go:74\nruntime.gopanic\n\t/home/travis/.gimme/versions/go1.12.1.linux.amd64/src/runtime/panic.go:522\nruntime.panicmem\n\t/home/travis/.gimme/versions/go1.12.1.linux.amd64/src/runtime/panic.go:82\nruntime.sigpanic\n\t/home/travis/.gimme/versions/go1.12.1.linux.amd64/src/runtime/signal_unix.go:390\ngithub.com/jaegertracing/jaeger/model/adjuster.(*spanIDDeduper).groupSpansByID\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/model/adjuster/span_id_deduper.go:57\ngithub.com/jaegertracing/jaeger/model/adjuster.SpanIDDeduper.func1\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/model/adjuster/span_id_deduper.go:35\ngithub.com/jaegertracing/jaeger/model/adjuster.Func.Adjust\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/model/adjuster/adjuster.go:36\ngithub.com/jaegertracing/jaeger/model/adjuster.sequence.Adjust\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/model/adjuster/adjuster.go:62\ngithub.com/jaegertracing/jaeger/cmd/query/app/querysvc.QueryService.Adjust\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/query/app/querysvc/query_service.go:113\ngithub.com/jaegertracing/jaeger/cmd/query/app.(*APIHandler).convertModelToUI\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/query/app/http_handler.go:279\ngithub.com/jaegertracing/jaeger/cmd/query/app.(*APIHandler).getTrace\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/query/app/http_handler.go:360\nnet/http.HandlerFunc.ServeHTTP\n\t/home/travis/.gimme/versions/go1.12.1.linux.amd64/src/net/http/server.go:1995\ngithub.com/jaegertracing/jaeger/vendor/github.com/opentracing-contrib/go-stdlib/nethttp.MiddlewareFunc.func5\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/vendor/github.com/opentracing-contrib/go-stdlib/nethttp/server.go:132\nnet/http.HandlerFunc.ServeHTTP\n\t/home/travis/.gimme/versions/go1.12.1.linux.amd64/src/net/http/server.go:1995\nnet/http.HandlerFunc.ServeHTTP\n\t/home/travis/.gimme/versions/go1.12.1.linux.amd64/src/net/http/server.go:1995\ngithub.com/jaegertracing/jaeger/vendor/github.com/gorilla/mux.(*Router).ServeHTTP\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/vendor/github.com/gorilla/mux/mux.go:212\ngithub.com/jaegertracing/jaeger/vendor/github.com/gorilla/handlers.CompressHandlerLevel.func1\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/vendor/github

It crashes the system and then takes a while to restart badger?

**Comment by burmanm** [comment]
There's nothing in that stacktrace indicating the badger backend causes the nil pointer though (it's not anywhere in that path).

**Comment by burmanm** [comment]
What version are you running? In any case, it looks like in the span_deduper the returned span is without traces:

```go
for _, span := range d.trace.Spans {
```

because that line causes the issue. So the d.trace.Spans == nil (which isn't checked..)

**Comment by JoseAlban** [comment]
I'm on jaeger-collector 1.14.0 and got similar problem on the same line:

```
{"level":"error","ts":1572431197.4916267,"caller":"recoveryhandler/zap.go:33","msg":"runtime error: index out of range","stacktrace":"github.com/jaegertracing/jaeger/pkg/recoveryhandler.zapRecoveryWrapper.Println\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/pkg/recoveryhandler/zap.go:33\ngithub.com/jaegertracing/jaeger/vendor/github.com/gorilla/handlers.recoveryHandler.log\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/vendor/github.com/gorilla/handlers/recovery.go:83\ngithub.com/jaegertracing/jaeger/vendor/github.com/gorilla/handlers.recoveryHandler.ServeHTTP.func1\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/vendor/github.com/gorilla/handlers/recovery.go:74\nruntime.gopanic\n\t/home/travis/.gimme/versions/go1.12.1.linux.amd64/src/runtime/panic.go:522\nruntime.panicindex\n\t/home/travis/.gimme/versions/go1.12.1.linux.amd64/src/runtime/panic.go:44\ngithub.com/jaegertracing/jaeger/cmd/collector/app.(*zipkinSpanHandler).SubmitZipkinBatch\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/collector/app/thrift_span_handler.go:126\ngithub.com/jaegertracing/jaeger/cmd/collector/app/zipkin.(*APIHandler).saveThriftSpans\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/collector/app/zipkin/http_handler.go:204\ngithub.com/jaegertracing/jaeger/cmd/collector/app/zipkin.(*APIHandler).saveSpans\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/collector/app/zipkin/http_handler.go:106\nnet/http.HandlerFunc.ServeHTTP\n\t/home/travis/.gimme/versions/go1.12.1.linux.amd64/src/net/http/server.go:1995\ngithub.com/jaegertracing/jaeger/vendor/github.com/gorilla/mux.(*Router).ServeHTTP\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/vendor/github.com/gorilla/mux/mux.go:212\ngithub.com/jaegertracing/jaeger/vendor/github.com/gorilla/handlers.recoveryHandler.ServeHTTP\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/vendor/github.com/gorilla/handlers/recovery.go:78\ngithub.com/jaegertracing/jaeger/vendor/github.com/rs/cors.(*Cors).Handler.func1\n\t/home/travis/gopath/src/github.com/jaegertracing/jaeger/vendor/github.com/rs/cors/cors.go:207\nnet/http.HandlerFunc.ServeHTTP\n\t/home/travis/.gimme/versions/go1.12.1.linux.amd64/src/net/http/server.go:1995\nnet/http.serverHandler.ServeHTTP\n\t/home/travis/.gimme/versions/go1.12.1.linux.amd64/src/net/http/server.go:2774\nnet/http.(*conn).serve\n\t/home/travis/.gimme/versions/go1.12.1.linux.amd64/src/net/http/server.go:1878"}
```

**Comment by IvanBodnya** [comment]
Having same issue with span_deduper. Jaeger crashes on the same line as described above.
Any news on when this issue can be resolved?

**Comment by jpkrohling** [comment]
Would you like to give it a try, @IvanBodnya?

**Comment by IvanBodnya** [comment]
Yeah, sure. I've already forked the repo. I'm new to golang, but I'll definitely go for it!
